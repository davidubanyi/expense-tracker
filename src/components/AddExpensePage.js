import React from "react";

const AddExpensePage = () => {
  return <div>this is the create page</div>;
};

export default AddExpensePage;
