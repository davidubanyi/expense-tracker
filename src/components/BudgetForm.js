import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";

class BudgetForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      group: props.budget ? props.budget.group : "",
      description: props.budget ? props.budget.description : "",
      amount: props.budget ? (props.budget.amount / 100).toString() : "",
      createdAt: props.budget ? moment(props.budget.createdAt) : moment(),
      categories: props.categoryGroup,
      calenderFocused: false,
      error: ""
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({
      description
    }));
  };

  onGroupChange = e => {
    const group = e.target.value.toLowerCase();
    this.setState(() => ({
      group
    }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount
      }));
    }
  };
  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({
        createdAt
      }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({
      calenderFocused: focused
    }));
  };

  onCategoryClick = e => {
    this.onGroupChange(e);
  };
  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please fill in the description and amount fields"
      }));
    } else {
      this.setState(() => ({
        error: ""
      }));
      this.props.onSubmit({
        group: this.state.group || "random",
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf()
      });
      if (
        this.state.group &&
        !this.state.categories.includes(this.state.group)
      ) {
        this.props.addCategory(this.state.group);
      }
      this.setState({
        group: "",
        description: "",
        amount: ""
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        {this.state.categories.map(category => {
          return (
            <button
              key={category}
              value={category}
              onClick={this.onCategoryClick}
            >
              {category}
            </button>
          );
        })}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Spending group e.g food"
            value={this.state.group}
            onChange={this.onGroupChange}
          />
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => {
              false;
            }}
          />
          <button>Add Budget</button>
        </form>
      </div>
    );
  }
}

export default BudgetForm;
