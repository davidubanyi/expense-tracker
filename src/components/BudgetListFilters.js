import React from "react";
import { connect } from "react-redux";
import {
  setGroupFilter,
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate
} from "../redux/actions/filters";
import { DateRangePicker } from "react-dates";

export class BudgetListFilters extends React.Component {
  state = {
    calenderFocused: null,
    active: ""
  };

  componentDidMount(){
    this.props.setGroupFilter(this.state.active)
    this.props.setTextFilter(this.state.active)
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calenderFocused => {
    this.setState(() => ({
      calenderFocused
    }));
  };

  onTextInputChange = e => {
    this.props.setTextFilter(e.target.value);
  };

  onSelectInputChange = e => {
    if (e.target.value === "date") {
      this.props.sortByDate();
    } else if (e.target.value === "amount") {
      this.props.sortByAmount();
    }
  };

  onCategoryClick = e => {
    const category = e.target.value
    this.setState(()=>({
      active: category
    }))
    this.props.setGroupFilter(category);
  
  };

  render() {
    return (
      <div className="list-filters--container">
        <input
        id="search"
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextInputChange}
          placeholder="search"
        />
        <div className="button-tags">
          <button
            className={this.state.active === "" ? "active-button" : ""}
            onClick={this.onCategoryClick}
            value=""
          >
            all
          </button>
          {this.props.categories.map(category => {
            return (
              <button
                className={
                  this.state.active === category ? "active-button" : ""
                }
                key={category}
                value={category}
                onClick={this.onCategoryClick}
              >
                {category}
              </button>
            );
          })}
        </div>
        <select onChange={this.onSelectInputChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          startDateId="startDate"
          endDateId="endDate"
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calenderFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        
      </div>
    );
  }
}

const mapStateToProps = state => ({
  filters: state.filter,
  categories: state.category
});

const mapDispatchToProps = dispatch => ({
  setGroupFilter: groupName => dispatch(setGroupFilter(groupName)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate())
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(BudgetListFilters);
