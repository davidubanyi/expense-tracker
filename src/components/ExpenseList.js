import React from "react";
import { connect } from "react-redux";
import { NavLink } from "react-router-dom";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../redux/selectors/expenses";

export const ExpenseList = props => {
  return (
    <div>
      {props.expenses.length === 0 && !props.addExpense ? (
        <div>
          <p>No Expenses</p>
          <NavLink to="/add_expenses">Add Expense</NavLink>
        </div>
      ) : (
        props.expenses.map(expense => {
          return <ExpenseListItem key={expense.id} {...expense} />;
        })
      )}
    </div>
  );
};

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filter)
  };
};

export default connect(mapStateToProps)(ExpenseList);
