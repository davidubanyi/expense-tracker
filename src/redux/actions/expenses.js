import db from "../../firebase/firebase";
import uuid from 'uuid'

//ADD_EXPENSE
export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      group = "",
      description = "",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { group, description, note, amount, createdAt };
    const id = uuid()
    dispatch(addExpense({ id: id, ...expense }));
    db.collection("expenses")
      .doc(id).set(expense)
      .then(() => {
        console.log('added to db successfully')
      });
  };
};
//REMOVE_EXPENSE
export const removeExpense = id => ({
  type: "REMOVE_EXPENSE",
  id
});

export const startRemoveExpense = (id) => {
  return dispatch => {
    dispatch(removeExpense(id))
     db.collection("expenses").doc(id).delete().then(function(){
      console.log('deleted successfully from db')
    })
  }
}


//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return dispatch => {
    dispatch(editExpense(id, updates))
    db.collection("expenses").doc(id).update(updates).then(()=> {
      console.log('edited from the db successfully')
    })
  }
}

//SET EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

//Start Set Expenses
export const startSetExpenses = () => {
  return (dispatch) => {
    return db.collection('expenses').get().then(function(querySnapshot){
      const expenses = []
      querySnapshot.forEach((doc)=>{
        const data = doc.data()
        expenses.push({
          id: doc.id,
          ...data
        })
      })
      dispatch(setExpenses(expenses))
    })
  }
}