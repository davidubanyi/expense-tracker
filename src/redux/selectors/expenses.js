import moment from "moment";

//get visible expenses
export default (expenses, { text, sortBy, startDate, endDate, group }) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());
      const groupMatch = expense.group
        .toLowerCase()
        .includes(group.toLowerCase());

      return startDateMatch && endDateMatch && textMatch && groupMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};
