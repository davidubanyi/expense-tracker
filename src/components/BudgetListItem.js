import React from "react";
import { connect } from "react-redux";
import { startRemoveBudget } from "../redux/actions/budget";
import { Link } from "react-router-dom";
import moment from 'moment'
import {toast, Flip} from 'react-toastify'
import numeral from 'numeral'

export const BudgetListItem = ({
  dispatch,
  group,
  description,
  amount,
  createdAt,
  id
}) => {
  const notify = () =>
    toast.warn("Deleted Budget", {
      position: toast.POSITION.TOP_CENTER,
      transition: Flip
    });
  return (
    <div className="list-item--container">
      <div className="list-item--details">
        <p>{group}</p>
        <h2>{description}</h2>
        <span>{moment(createdAt).format("ddd, Do MMMM")}</span>
      </div>
      <div className="list-item--amount">
        <h1>{numeral(amount / 100).format('$0,0.00')}</h1>
        <button
          onClick={e => {
            notify()
            dispatch(startRemoveBudget(id));
          }}
        >
          Remove
        </button>
        <Link to={`/edit_budget/${id}`}>Edit</Link>
      </div>
    </div>
  );
};

export default connect()(BudgetListItem);
