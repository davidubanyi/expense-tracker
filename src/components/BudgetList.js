import React from "react";
import { connect } from "react-redux";
import BudgetListItem from "./BudgetListItem";
import selectBudget from "../redux/selectors/budgets";

export const ExpenseList = props => {
  return (
    <div>
      {props.budgets.length === 0 ? (
        <p>No Budget</p>
      ) : (
        props.budgets.map(budget => {
          return <BudgetListItem key={budget.id} {...budget} />;
        })
      )}
    </div>
  );
};

const mapStateToProps = state => {
  return {
    budgets: selectBudget(state.budget, state.filter)
  };
};

export default connect(mapStateToProps)(ExpenseList);
