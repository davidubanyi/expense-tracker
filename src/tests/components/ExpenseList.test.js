import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("should render ExpensesList with expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test("should render No Expenses when there is no expense", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
 