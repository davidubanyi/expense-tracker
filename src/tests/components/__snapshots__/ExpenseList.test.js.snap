// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesList with expenses 1`] = `
<div>
  <Connect(ExpenseListItem)
    amount={195}
    createdAt={0}
    description="gum"
    id="1"
    key="1"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={109500}
    createdAt={-345600000}
    description="rent"
    id="2"
    key="2"
    note=""
  />
  <Connect(ExpenseListItem)
    amount={9500}
    createdAt={345600000}
    description="machinery"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`should render No Expenses when there is no expense 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;
